src/core/manager.c

```diff
@@ -126,6 +127,107 @@ static int manager_run_environment_generators(Manager *m);
 static int manager_run_generators(Manager *m);
 static void manager_vacuum(Manager *m);
 
+int manager_spawn_workers(Manager *m) {
+        char socket_fd_number[DECIMAL_STR_MAX(int) + 1];
+        _cleanup_free_ char *log_level = NULL;
+        static int n_workers_cache = -2;
+        const char *scope_path;
+        size_t n_workers;
+        int r;
+
+        assert(m);
+
+        if (FLAGS_SET(m->test_run_flags, MANAGER_TEST_RUN_MINIMAL))
+                return 0;
+
+        if (n_workers_cache < -1) {
+                const char *e = secure_getenv("SYSTEMD_WORKERS_POOL_SIZE");
+                if (e) {
+                        r = safe_atoi(e, &n_workers_cache);
+                        if (r < 0) {
+                                log_debug("Invalid value in $SYSTEMD_WORKERS_POOL_SIZE, ignoring: %s", e);
+                                n_workers_cache = -1;
+                        }
+                } else
+                        n_workers_cache = -1;
+        }
+
+        if (n_workers_cache == 0)
+                return 0;
+
+        scope_path = strjoina(m->cgroup_root, "/" SPECIAL_INIT_WORKERS_SCOPE);
+
+        r = cg_is_empty(SYSTEMD_CGROUP_CONTROLLER, scope_path);
```

Comment:
hmm, maybe use pids.current cgroup attribute to determine how many processes are still left in the cgroup, and then simply top up? then we could call this function any time we like and we'd add what's missing, very robustly.

Reply:
I tried that first, but it required the pid controller to be enabled. Which is fine when running as the system manager, but as an unprivileged process especially in tests all sort of stuff broke apart, as the controller was not available in all incarnations of the CI, IIRC especially on CentOS there were issues. On the other hand this helper works everywhere and in all conditions, even when starting a manager in a unit test.

Reply:
maybe add a helper that falls back to counting processes simply via cg_enumerate_processes(). It's not quite the same (because processes != threads), but in our case should be OK, because the executor is single-threaded.

Reply:
I'm not sure I want to do a top up on every round around the loop, as that effectively means spawning one process every time a service is started. I think it works better to spawn a bunch in a tight loop and then move on.
Later we should do better than both, and save some stats on hits/misses and adjust the pool size dynamically based on that. But that's a future improvement.